docs/learn/multi-format-reader.md

```diff
@@ -0,0 +1,263 @@
+        """
+        ext = os.path.splitext(path)[1]
+        if self.processing_order is None or ext not in self.processing_order:
+            return (1, ext)
+        return (0, self.processing_order.index(ext))
+
+    sorted_paths = sorted(file_paths, key=get_processing_order)
+```
+If the reader has a `self.processing_order`, the input files get sorted in this order. If `self.overwrite_keys` is True, later files get precedent.
+
+### Reading of input files
+```python
+    for file_path in sorted_paths:
+        extension = os.path.splitext(file_path)[1].lower()
+        if extension not in self.extensions:
+            logger.warning(
+                f"File {file_path} has an unsupported extension, ignoring file."
+            )
+            continue
+        if not os.path.exists(file_path):
+            logger.warning(f"File {file_path} does not exist, ignoring entry.")
+            continue
+
+        template.update(self.extensions.get(extension, lambda _: {})(file_path))
+```
+This parts reads in the data from all data files. The `MultiFormatReader` has an `extensions` property, which is a dictionary that for each file extension calls a function that reads in data from files with that extension. If the reader shall handle e.g. an HDF5 file, a method for handling this type of file should be added, i.e., `self.extensions[".hdf5"] = self.handle_hdf5`. 
+Note that these method should also implement any case selection logic, i.e., it may not be sufficient to rely on the filename suffix, but when may also need to check for multiple versions, binary signature, mimetype, etc.
+
+Any of these methods should take as input only the file path, e.g.
+```python
+def handle_eln_file(self, file_path: str) -> Dict[str, Any]
+```
+These methods must return a dictionary. One possibility is to return a dictionary that directly fills the template (see the `template.update` call above) with the data from the file. Another option is to return an empty dictionary (i.e., not fill the template at this stage) and only later fill the template from a config file (see below).
+
+Note that for several input formats, standardized parser functions already exist within the `MultiFormatReader`. For example, YAML files can be parsed using the `pynxtools.dataconverter.readers.utils.parse_yml` function.
+
+### Setting default values in the template
+```python
+    template.update(self.setup_template())
+```
+Next, the `setup_template` method can be implemented, which is used to populate the template with initial data that does not come from the files themselves. This may be used to set fixed information, e.g., about the reader. As an example, `NXentry/program_name` (which is defined as the name of program used to generate the NeXus file) scan be set to `pynxtools-plugin` by making `setup_template` return a dictionary of the form
+```json
+{
+  "/ENTRY[my_entry]/program_name": "pynxtools-plugin",
+  "/ENTRY[my_entry]/program_name/@version": "v0.1.0"
+}
+```

```

Comment:
Shoudn't this come before reading the files?

Reply:
Currently, it comes afterwards (and after potential hard-coded update of the templates in the reading functions). We can discuss to change the order if you like.

Reply:
I don't use either option, so in general I don't mind. Let's keep as is at least for now.